service/round_test.go

```diff
@@ -2,227 +2,322 @@ package service
 
 import (
 	"context"
-	"crypto/rand"
-	"fmt"
-	"path/filepath"
+	"math/rand"
+	"os"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/require"
 
+	"github.com/spacemeshos/poet/hash"
 	"github.com/spacemeshos/poet/prover"
+	"github.com/spacemeshos/poet/shared"
+	"github.com/spacemeshos/poet/verifier"
 )
 
+func genChallenge() ([]byte, error) {
+	challenge := make([]byte, 32)
+	_, err := rand.Read(challenge)
+	return challenge, err
+}
+
 func genChallenges(num int) ([][]byte, error) {
 	ch := make([][]byte, num)
 	for i := 0; i < num; i++ {
-		ch[i] = make([]byte, 32)
-		_, err := rand.Read(ch[i])
+		challenge, err := genChallenge()
 		if err != nil {
 			return nil, err
 		}
+		ch[i] = challenge
 	}
 
 	return ch, nil
 }
 
-// TestRound_Recovery test round recovery functionality.
-// The scenario proceeds as follows:
-//   - Execute r1 as a reference round.
-//   - Execute r2, and request shutdown before completion.
-//   - Recover r2 execution, and request shutdown before completion.
-//   - Recover r2 execution again, and let it complete.
-func TestRound_Recovery(t *testing.T) {
-	req := require.New(t)
-
-	ctx, stop := context.WithCancel(context.Background())
-	defer stop()
-	duration := 500 * time.Millisecond
-	tmpdir := t.TempDir()
+func numChallenges(r *round) int {
+	iter := r.challengesDb.NewIterator(nil, nil)
+	defer iter.Release()
 
-	challenges, err := genChallenges(32)
-	req.NoError(err)
+	var num int
+	for iter.Next() {
+		num++
+	}
 
-	// Execute r1 as a reference round.
-	r1, err := newRound(tmpdir, 0)
-	req.NoError(err)
-	req.NoError(r1.open())
-	req.Equal(0, r1.numChallenges())
-	req.True(r1.isEmpty())
+	return num
+}
 
-	for _, ch := range challenges {
-		req.NoError(r1.submit(ch, ch))
-	}
-	req.Equal(len(challenges), r1.numChallenges())
-	req.False(r1.isEmpty())
+func newTestRound(t *testing.T) *round {
+	t.Helper()
+	round, err := newRound(t.TempDir(), rand.Uint32())
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, round.teardown(true))
+	})
+	return round
+}
 
-	req.NoError(r1.execute(ctx, time.Now().Add(duration), prover.LowestMerkleMinMemoryLayer))
-	req.NoError(r1.teardown(true))
+// validateProof validates proof from round's execution state.
+func validateProof(t *testing.T, execution *executionState) {
+	t.Helper()
+	req := require.New(t)
+	req.NotEmpty(execution.Statement)
+	req.NotNil(execution.NIP)
+	req.NotEmpty(execution.NIP.Root)
+	req.NotEmpty(execution.NIP.ProvenLeaves)
+	req.NotEmpty(execution.NIP.ProofNodes)
+
+	req.NoError(
+		verifier.Validate(
+			*execution.NIP,
+			hash.GenLabelHashFunc(execution.Statement),
+			hash.GenMerkleHashFunc(execution.Statement),
+			execution.NumLeaves,
+			shared.T,
+		),
+	)
```

Comment:
Can this be shortened to just the last assertion? I assume `verifier.Validate` returns an error if any of the previous assertions would fail.